rocky/tools/forms/upload_raw.py

```diff
@@ -22,7 +30,60 @@ class UploadRawForm(forms.Form):
         allow_empty_file=False,
     )
 
+    ooi_id = forms.CharField(
+        label="Input or Scan OOI",
+        widget=DataListInput(
+            attrs={"placeholder": _("Click to select one of the available options, or type one yourself")}
+        ),
+    )
+
+    date = forms.DateTimeField(
+        label=_("Date/Time (UTC)"),
+        widget=DateTimeInput(format="%Y-%m-%dT%H:%M"),
+        help_text=RAW_FILE_DATETIME_HELP_TEXT,
+    )
+
+    def __init__(
+        self,
+        connector: OctopoesAPIConnector,
+        ooi_list: list[tuple[str, str]],
+        *args,
+        **kwargs,
+    ):
+        observed_at = kwargs.pop("observed_at", None)
+        super().__init__(*args, **kwargs)
+        self.octopoes_connector = connector
+        self.set_choices_for_widget("ooi_id", ooi_list)
+
+        if observed_at:
+            try:
+                parsed_date = (
+                    datetime.strptime(observed_at, "%Y-%m-%d").replace(tzinfo=timezone.utc).replace(hour=23, minute=59)
+                )
+                self.fields["date"].initial = parsed_date
+            except ValueError:
+                self.fields["date"].initial = datetime.now(tz=timezone.utc)
+        else:
+            self.fields["date"].initial = datetime.now(tz=timezone.utc)
+
     def clean_mime_types(self) -> set[str]:
         mime_types = self.cleaned_data["mime_types"]
 
         return {mime_type.strip() for mime_type in mime_types.split(",") if mime_type.strip()}
+
+    def clean(self):
+        cleaned_data = super().clean()
+
+        date = self.cleaned_data["date"]
+        ooi_id = self.cleaned_data["ooi_id"]
+
+        # date should not be in the future
+        if date > datetime.now(tz=timezone.utc):
+            self.add_error("date", _("Doc! I'm from the future, I'm here to take you back!"))
+
+        try:
+            cleaned_data["ooi"] = self.octopoes_connector.get(Reference.from_str(ooi_id), date)
```

Comment:
We allow, if I'm not mistaken, users to add raw files attached to OOIs no matter what scan level the OOI or user. Would that be an issue?

Reply:
If they have scanned out the scope of KAT, its out of our hands imho.